### PR TITLE
Initial crate

### PR DESCRIPTION
**Pull request by riverar**
The beginnings of the `windows-app` crate. Includes the wiring necessary to bootstrap dynamic dependencies for unpackaged Windows Apps and a native window handle helper.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+[package]
+name = "windows-app"
+version = "0.2.0"
+edition = "2018"
```
Comment (kennykerr) [inline Cargo.toml:4]: Here's what you'll need for publishing.

```suggestion
[package]
name = "windows-app"
version = "0.1.0"
authors = ["Microsoft"]
edition = "2018"
license = "MIT OR Apache-2.0"
description = "Windows App SDK"
repository = "https://github.com/microsoft/windows-app-rs"
documentation = "https://docs.rs/windows-app"
readme = "README.md"
exclude = [".github", ".windows", "docs"]
```

File: bindings/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "windows_app_bindings"
+version = "0.2.0"
+edition = "2018"
```
Comment (kennykerr) [inline bindings/Cargo.toml:4]:
```suggestion
[package]
name = "windows_app_bindings"
version = "0.2.0"
authors = ["Microsoft"]
edition = "2018"
license = "MIT OR Apache-2.0"
description = "Implementation for windows-app crate"
```

File: bindings/build.rs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+fn main() {
+    windows::build! {
+        Microsoft::Foundation::*,
+        Microsoft::UI::Xaml::Window,
+        Windows::Win32::Foundation::{BOOL, RECT},
+        Windows::Win32::UI::HiDpi::GetDpiForWindow,
+        Windows::Win32::UI::WindowsAndMessaging::*,
```
Comment (kennykerr) [inline bindings/build.rs:7]: Let's be very specific here to reduce build time.

File: build.rs [hunk 4]
```diff
@@ -0,0 +1,18 @@
+use std::{env, fs, path};
+
+fn main() {
```
Comment (kennykerr) [inline build.rs:3]: I don't see anything here that the build macro won't already do automatically if you simply place the files in the various `/.windows/*` folders.

**Comment by kennykerr** [comment]
Need a `.gitattributes` with the following to avoid having the binaries scrambled:

```
*       text=auto
*       text eol=lf
*.png   -text
*.winmd -text
*.dll   -text
```

**Comment by kennykerr** [comment]
Should also add a basic CI build at `/.github/workflows/build.yml` - this should suffice:

```
name: Build and Test

on:
  pull_request:
  push:
    branches:
      - master

env:
  RUSTFLAGS: -Dwarnings

jobs:
  test:
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        include:
        - os: windows-latest
          rust: stable
          other: x86_64-pc-windows-msvc
        - os: windows-latest
          rust: 1.51.0
          other: x86_64-pc-windows-msvc
        - os: windows-latest
          rust: nightly
          other: x86_64-pc-windows-msvc
        - os: windows-latest
          rust: stable
          other: i686-pc-windows-msvc
        - os: windows-latest
          rust: nightly
          other: i686-pc-windows-msvc
    steps:
    - uses: actions/checkout@v2
    - run: rustup update --no-self-update ${{ matrix.rust }} && rustup default ${{ matrix.rust }}
    - run: rustup target add ${{ matrix.other }}

    - name: windows build
      run: cargo test --all --target ${{ matrix.other }}

  cargo_fmt:
    runs-on: windows-latest
    name: Check cargo formatting
    steps:
    - uses: actions/checkout@v2
    - name: Run cargo fmt
      run: cargo fmt --all -- --check

  cargo_doc:
    runs-on: windows-latest
    name: Check cargo docs
    steps:
    - uses: actions/checkout@v2
    - name: Run cargo doc
      run: cargo doc --no-deps --workspace
```

File: src/bootstrap/workarounds.rs [hunk 5]
```diff
@@ -0,0 +1,112 @@
+use bindings::{
```
Comment (kennykerr) [inline src/bootstrap/workarounds.rs:1]: I'd like to avoid supporting these helpers unless there's an urgent blocking issue they resolve.

Comment (riverar) [reply src/bootstrap/workarounds.rs:1]: I tried to signal that via the `workarounds` namespace. Want to drop them altogether or rename? Every app developer will need these helpers due to gaps in WinUI.

Comment (kennykerr) [reply src/bootstrap/workarounds.rs:1]: Then perhaps just a `windows_app::window_handle(windows::IInspectable) -> isize` interop function and let developers do the rest by calling whatever APIs they need. Hopefully WinUI fills their gaps but at least then developers can plug whatever else they need in the meantime without us having to add wrappers for everything.

File: .gitattributes [hunk 6]
```diff
@@ -0,0 +1,5 @@
+*       text=auto
+*       text eol=lf
+*.png   -text
+*.winmd -text
+*.dll   -text
```
Comment (kennykerr) [inline .gitattributes:5]: need to add .lib

File: bindings/build.rs [hunk 7]
```diff
@@ -0,0 +1,11 @@
+fn main() {
+    windows::build! {
+        Microsoft::UI::Xaml::Window,
```
Comment (kennykerr) [inline bindings/build.rs:3]: Probably don't need much more than `MessageBoxW` here.

File: src/bootstrap/mod.rs [hunk 8]
```diff
@@ -0,0 +1,88 @@
+use bindings::Windows::Win32::{
+    Foundation::HWND,
+    UI::WindowsAndMessaging::{MessageBoxW, MB_ICONERROR, MB_OK},
+};
+
+mod tests;
```
Comment (rylev) [inline src/bootstrap/mod.rs:6]: Unit tests are normally in the same file (in a submodule). This is fine, but perhaps not the most idiomatic.

File: src/bootstrap/mod.rs [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+                "This application could not be started",
+                MB_OK | MB_ICONERROR,
+            );
+            outer_error
+        }
+    })
+}
+
+pub fn initialize_without_dialog() -> windows::Result<()> {
+    let version_tag: Vec<u16> = "preview".encode_utf16().collect();
```
Comment (rylev) [inline src/bootstrap/mod.rs:62]: Taking a dependency on https://crates.io/crates/const-utf16 or some equivalent might be helpful.

File: src/bootstrap/mod.rs [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    unsafe {
```
Comment (rylev) [inline src/bootstrap/mod.rs:63]: I think it’s best practice to limit the unsafe blocks to only code that needs it. Creating the PackageVersion structs is not unsafe and so it doesn’t need to be wrapped in an unsafe block.

**Review by rylev — APPROVED** [review]
Excited to see where this goes!

File: src/bootstrap/mod.rs (hunk 9 above)
Comment (kennykerr) [reply src/bootstrap/mod.rs:62]: This should soon be baked in, so perhaps we can avoid the extra dependency for the short term.

**Comment by riverar** [comment]
Oops @rylev just saw your review as I hit push. Reviewing now.

**Comment by kennykerr** [comment]
woohoo - CI build works!

File: src/bootstrap/mod.rs (hunk 9 above)
Comment (riverar) [reply src/bootstrap/mod.rs:62]: Yep, eventually we'll pass this requirement onto the developer; they must provide the correct version of the dynamic dependency component. Still in talks with the Windows App SDK folks to figure out exactly how all this is going to work, so went with _working values_ for now.

**Review by kennykerr — APPROVED** [review]

**Pull request merged**